src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunction.Activity.cs

```diff
@@ -179,14 +181,44 @@ public async Task ReindexBatchV2Async([ActivityTrigger] ReindexBatchArguments ar
         IReadOnlyList<VersionedInstanceIdentifier> instanceIdentifiers =
             await _instanceStore.GetInstanceIdentifiersByWatermarkRangeAsync(arguments.WatermarkRange, IndexStatus.Created);
 
+        var missing = new ConcurrentDictionary<VersionedInstanceIdentifier, ItemNotFoundException>();
+
         await Parallel.ForEachAsync(
             instanceIdentifiers,
             new ParallelOptions
             {
                 CancellationToken = default,
                 MaxDegreeOfParallelism = _options.MaxParallelThreads,
             },
-            (id, token) => new ValueTask(_instanceReindexer.ReindexInstanceAsync(arguments.QueryTags, id, token)));
+            async (id, token) =>
+            {
+                // Record any failed re-indexing
+                try
+                {
+                    await _instanceReindexer.ReindexInstanceAsync(arguments.QueryTags, id, token);
+                }
+                catch (ItemNotFoundException ex)
+                {
+                    missing[id] = ex;
+                }
+            });
+
+        // If there are any missing, re-query the batch and ensure they were deleted
+        if (!missing.IsEmpty)
+        {
+            logger.LogWarning("Could not find {Count} instances to re-index", arguments.WatermarkRange, missing.Count);
+            instanceIdentifiers = await _instanceStore.GetInstanceIdentifiersByWatermarkRangeAsync(arguments.WatermarkRange, IndexStatus.Created);
+
+            foreach (VersionedInstanceIdentifier identifier in instanceIdentifiers)
+            {
+                // If the identifier is still present, then throw!
+                if (missing.TryGetValue(identifier, out ItemNotFoundException exception))
+                {
+                    logger.LogError("Cannot find metadata for watermark {Watermark}", identifier.Version);
```

Comment:
It is either something that's transitory (a delete + an undo), or something that's a problem on our end that we will have to investigate in the logs. To the user, if it fails after enough tries, it will be manifest as a failed operation

Reply:
As discussed, there is currently no TSG, but if this specific scenario happens, then we can create one. More generally, we would investigate any failed orchestration, like extended query tags or export.